util/yaml.go

```diff
@@ -37,3 +38,33 @@ func UnmarshalYAMLStrictWithFallback(in []byte, out interface{}) error {
 	}
 	return nil
 }
+
+// UnmarshalUpgradedYAMLWithFallback is identical to UnmarshalYAMLWithFallback but uses the upgraded YAML version
+// (v3.0.1) instead of the current YAML version.
+// This should only be used for user validation to see if their project's YAML hypothetically produces a valid config
+// should the YAML version be upgraded.
+func UnmarshalUpgradedYAMLWithFallback(in []byte, out interface{}) error {
+	if err := upgradedYAML.Unmarshal(in, out); err != nil {
+		// try the older version of yaml before erroring, in case it's just an outdated yaml
+		if err2 := yaml2.Unmarshal(in, out); err2 != nil {
```

Comment:
I'm thinking we shouldn't use fallback here at all; the whole purpose of fallback was to avoid having to force users to update things to use yaml3 since lots of projects had issues with it, but if we allow falling back to yaml2 in the upgraded case then we _will_ end up hiding the errors people are explicitly looking for with the flag, since the nested issue probably won't fail for yaml2.

Reply:
The fallback is only necessary if their YAML is not already v3 compatible, which is not affected by the YAML upgrade. If it's not v3 compatible, it will continue being incompatible and parse with v2 fallback logic (i.e. the v3 YAML upgrade is completely irrelevant to them). If it _is_ v3 compatible but is subject to the nested map bug, it will not error, but will instead produce a different (but valid) parsed result, which can only be checked by later logic.